mesop/dataclass_utils/dataclass_utils.py

```diff
@@ -127,3 +139,73 @@ def decode_mesop_json_state_hook(dct):
   except ImportError:
     pass
   return dct
+
+
+class DataFrameOperator(BaseOperator):
+  """Custom operator to detect changes in DataFrames.
+
+  DeepDiff does not support diffing DataFrames. See https://github.com/seperman/deepdiff/issues/394.
+
+  This operator checks if the DataFrames are equal or not. It does not do a deep diff of
+  the contents of the DataFrame.
+  """
+
+  def match(self, level) -> bool:
+    try:
+      import pandas as pd
+
+      return isinstance(level.t1, pd.DataFrame) and isinstance(
+        level.t2, pd.DataFrame
+      )
+    except ImportError:
+      # If Pandas is not installed, don't perform this check. We should log a warning.
+      return False
+
+  def give_up_diffing(self, level, diff_instance) -> bool:
+    if not level.t1.equals(level.t2):
+      diff_instance.custom_report_result(
+        _DIFF_ACTION_DATA_FRAME_CHANGED, level, {"value": level.t2}
+      )
+    return True
+
+
+def diff_state(state1: Any, state2: Any) -> str:
+  """
+  Diffs two state objects and returns the difference using DeepDiff's Delta format as a
+  JSON string.
+
+  DeepDiff does not support DataFrames yet. See `DataFrameOperator`.
+
+  The `to_flat_dicts` method does not include custom report results, so we need to add
+  those manually for the DataFrame case.
+  """
+  if not is_dataclass(state1) or not is_dataclass(state2):
+    raise MesopException("Tried to diff state which was not a dataclass")
+
+  custom_actions = []
+
+  # Only use the `DataFrameOperator` if the state class has DataFrames since we don't
+  # want to import Pandas unnecessarily.
+  if "DataFrame" in str(state1.__class__.__annotations__):
```

Comment:
If the DataFrame is in a nested class (e.g. dataclass) inside a state class, Will this work?

Reply:
Ah good catch. Won't work. I updated to check for pandas globally once. Not sure if it makes difference from just repeatedly checking if the import exists.

Reply:
repeatedly checking is probably fine b/c Python will cache after the first import, but this is fine too.